File: December 2024 Leetcode Solution/December-07.py
class Solution:
    def minimumSize(self, nums: List[int], maxOperations: int) -> int:
        def canDivide(mx: int) -> bool:
            operations = 0
            for x in nums:
                operations += (x - 1) // mx
                if operations > maxOperations:
                    return False
            return True

        left, right = 1, max(nums)
        while left < right:
            mid = (left + right) // 2
            if canDivide(mid):
                right = mid
            else:
                left = mid + 1
        return left

class Solution:
    def minimumSize(self, nums: List[int], maxOperations: int) -> int:
        def canDivide(maxSize: int) -> bool:
            operations = sum((num - 1) // maxSize for num in nums)
            return operations <= maxOperations

        left, right = 1, max(nums)
        while left < right:
            mid = (left + right) // 2
            if canDivide(mid):
                right = mid
            else:
                left = mid + 1
        return left

class Solution:
    def minimumSize(self, nums: List[int], maxOperations: int) -> int:
        def canDivide(maxSize: int) -> bool:
            return sum((num - 1) // maxSize for num in nums) <= maxOperations

        left, right = 1, max(nums)
        while left < right:
            mid = left + (right - left) // 2
            if canDivide(mid):
                right = mid  
            else:
                left = mid + 1  
        return left

File: December 2024 Leetcode Solution/December-07.rs
impl Solution {
    pub fn minimum_size(nums: Vec<i32>, max_operations: i32) -> i32 {
        fn can_divide(nums: &Vec<i32>, max_size: i32, max_operations: i32) -> bool {
            nums.iter()
                .map(|&num| (num - 1) / max_size)
                .sum::<i32>() <= max_operations
        }

        let (mut left, mut right) = (1, *nums.iter().max().unwrap());

        while left < right {
            let mid = left + (right - left) / 2;
            if can_divide(&nums, mid, max_operations) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        left
    }
}
